object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java

```diff
@@ -338,44 +362,153 @@ private void handleAssignment(AssignmentNode node, Set<LocalVarWithTree> newDefs
+          LocalVariableNode n = getMustCallChoiceParam(rhs);
+          if (n != null && isVarInDefs(newDefs, n)) {
+            ImmutableSet<LocalVarWithTree> immutableSet =
+                getSetContainingAssignmentTreeOfVar(newDefs, n);
+            ImmutableSet<LocalVarWithTree> newImmutableSet =
+                FluentIterable.from(immutableSet)
+                    .append(
+                        new LocalVarWithTree(
+                            new LocalVariable((LocalVariableNode) lhs), node.getTree()))
+                    .toSet();
+            newDefs = updateDefs(newDefs, immutableSet, newImmutableSet);
+          }
+        } else {
+          LocalVarWithTree lhsLocalVarWithTreeNew =
+              new LocalVarWithTree(new LocalVariable((LocalVariableNode) lhs), node.getTree());
+          newDefs.add(ImmutableSet.of(lhsLocalVarWithTreeNew));
+        }
       }
 
       // Ownership Transfer
       if (rhs instanceof LocalVariableNode && isVarInDefs(newDefs, (LocalVariableNode) rhs)) {
         // If the rhs is a LocalVariableNode that exists in the newDefs (Note that if a
         // localVariableNode exists in the newDefs it means it isn't assigned to a null
         // literals), then it adds the localVariableNode to the newDefs
-        newDefs.add(
-            new LocalVarWithTree(new LocalVariable((LocalVariableNode) lhs), node.getTree()));
-        newDefs.remove(getAssignmentTreeOfVar(newDefs, (LocalVariableNode) rhs));
+        ImmutableSet<LocalVarWithTree> setContainingRhs =
+            getSetContainingAssignmentTreeOfVar(newDefs, (LocalVariableNode) rhs);
+        LocalVarWithTree lhsLocalVarWithTreeNew =
+            new LocalVarWithTree(new LocalVariable((LocalVariableNode) lhs), node.getTree());
+        newDefs.add(ImmutableSet.of(lhsLocalVarWithTreeNew));
+        newDefs.remove(setContainingRhs);
+      }
+    }
+    return newDefs;
+  }
+
+  private Set<ImmutableSet<LocalVarWithTree>> updateDefs(
+      Set<ImmutableSet<LocalVarWithTree>> defs,
+      ImmutableSet<LocalVarWithTree> oldSet,
+      ImmutableSet<LocalVarWithTree> newSet) {
+    return defs.stream().map(set -> set.equals(oldSet) ? newSet : set).collect(toSet());
+  }
+
+  /**
+   * Finds the {@code @MustCallChoice} parameter. If {@code node} represents a method invocation or
+   * object creation node: 1) If there is a formal parameter with {@link @MustCallChoice}
+   * annotation, it passes the parameter to a recursive call to handle nested calls 2) If there is
+   * no formal parameter with {@link @MustCallChoice} annotation, then it checks the receiver
+   * parameter. If the receiver parameter has @MustCallChoice annotation, then it passes the
+   * receiver parameter to a recursive call to handle sequence of method calls
+   *
+   * @param node
+   * @return {@code node} iff {@code node} represents a local variable, otherwise null
+   */
+  private @Nullable LocalVariableNode getMustCallChoiceParam(Node node) {
```

Comment:
Let's rename the method to `getLocalPassedAsMustCallChoiceParam()`.  The documentation needs to be improved.  Basically, this method tries to find a local variable passed as a `@MustCallChoice` parameter.  In the base case, if `node` is a local variable, it just gets returned.  Otherwise, if `node` is a call (or a call wrapped in a cast), the code finds the parameter passed in the `@MustCallChoice` position, and recurses on that parameter.  Let's rewrite the doc to be more like this, and leave out the details about handling receivers